lib/recursive-methods.rb

```diff
@@ -1,49 +1,107 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: ? O(n)
+# Space complexity: ? O(1)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    # raise NotImplementedError, "Method not implemented"
+    t_fact(n, 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def t_fact(n, i)
+    if n < 0
+        raise ArgumentError, "number is less than zero"
+    else
+        return i if n == 1 || n == 0
+    end
+
+    return t_fact(n - 1, i * n)
+end
+
+# Time complexity: ?O(n)
+# Space complexity: ? O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    # raise NotImplementedError, "Method not implemented"
+    return s if s.length <= 1
+
+    return s[-1] + reverse(s[0..-2])
+
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: ? O(n)
+# Space complexity: ? O(n^2)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+
+    inplace_helper(s, 0)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+
+def inplace_helper(s, i)
+
+    return s if i == s.length / 2
+    s[0 + i], s[-1 - i] = s[-1 - i], s[0 + i]
+
+    inplace_helper( s, i + 1)
+end
+
+# Time complexity: ? O(n)
+# Space complexity: ? O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    return 0 if n == 0
+
+    if n == 1
+        return 2
+    else
+        return 2 + bunny(n - 1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: ? O(n^2)
+# Space complexity: ? O(n^2)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length < 1
+        return true
+    elsif s[0] == s[-1]
+        return false
+    else
+        return nested(s[1..-2])
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: ? O(n^2)
+# Space complexity: ? O(n^2)
 def search(array, value)
```

Comment:
👍 , however can you see a better way to do this with O(n) space/time complexity?